llvm/lib/CheerpUtils/SIMDLowering.cpp

```diff
@@ -0,0 +1,284 @@
+	if (elementType->isIntegerTy())
+		amount = 128 / elementType->getIntegerBitWidth();
+	else if (elementType->isFloatTy())
+		amount = 4;
+	else if (elementType->isDoubleTy())
+		amount = 2;
+	else
+		llvm::report_fatal_error("Unknown elementwidth");
+	Value* newVec;
+	for (int i = 0; i < amount; i++)
+	{
+		Value* subOp1 = Builder.CreateExtractElement(firstOp, i);
+		Value* subOp2 = Builder.CreateExtractElement(secondOp, i);
+		Value* shiftedValue = Builder.CreateBinOp((Instruction::BinaryOps)opcode, subOp1, subOp2);
+		if (i == 0)
+		{
+			std::vector<Type *> argTypes { secondOp->getType(), elementType };
+			Function* splatIntrinsic = Intrinsic::getDeclaration(I.getModule(), Intrinsic::cheerp_wasm_splat, argTypes);
+			newVec = Builder.CreateCall(splatIntrinsic, { shiftedValue });
+		}
+		else
+			newVec = Builder.CreateInsertElement(newVec, shiftedValue, i);
+	}
+	I.replaceAllUsesWith(newVec);
+	deleteList.push_back(&I);
+	return false;
+}
+
+bool SIMDLoweringPass::isVariableExtractOrInsert(Instruction& I)
+{
+	return (I.getOpcode() == Instruction::ExtractElement && !isa<ConstantInt>(I.getOperand(1))) || 
+			(I.getOpcode() == Instruction::InsertElement && !isa<ConstantInt>(I.getOperand(2)));
+}
+
+bool SIMDLoweringPass::isReduceIntrinsic(Instruction& I)
+{
+	if (I.getOpcode() != Instruction::Call)
+		return false;
+	Intrinsic::ID id = cast<CallInst>(I).getCalledFunction()->getIntrinsicID();
+	return (id == Intrinsic::vector_reduce_mul ||
+		id == Intrinsic::vector_reduce_add ||
+		id == Intrinsic::vector_reduce_fmul ||
+		id == Intrinsic::vector_reduce_fadd);
+}
+
+PreservedAnalyses SIMDLoweringPass::run(Function& F, FunctionAnalysisManager& FAM)
+{
+	if (!WasmSIMD)
+		return PreservedAnalyses::all();
+	extractInsertAlloca = nullptr;
+	deleteList.clear();
+	bool needToBreak;
+	for (auto it = F.begin(); it != F.end(); it++)
+	{
+		BasicBlock& BB = *it;
+		for (Instruction& I: BB)
+		{
+			// Check correctness of vector types. Can only be 128 bits, or in some cases 64.
+#ifndef NDEBUG
+			if (I.getType()->isVectorTy())
+				checkVectorCorrectness(I);
+#endif
+			// This will find certain instructions that do not allow variables as lane indexes and
+			// instead use a store and a load.
+			if (isVariableExtractOrInsert(I))
+				needToBreak = lowerExtractOrInsert(I);
+			else if (isReduceIntrinsic(I))
+				needToBreak = lowerReduceIntrinsic(I);
+			else if ((I.getOpcode() == Instruction::Shl || I.getOpcode() == Instruction::AShr ||
+					I.getOpcode() == Instruction::LShr) && I.getType()->isVectorTy())
+				needToBreak = lowerBitShift(I);
+			else if (I.getOpcode() == Instruction::ShuffleVector)
+				needToBreak = lowerSplat(I);
+			else if ((I.getOpcode() == Instruction::SDiv || I.getOpcode() == Instruction::UDiv)
+					&& I.getType()->isVectorTy())
+				needToBreak = lowerGeneralUnsupportedVectorOperation(I);
+			if (needToBreak)
+				break ;
+		}
+	}
+	for (Instruction* I: deleteList)
+		I->eraseFromParent();
+	PreservedAnalyses PA;
```

Comment:
I would consider returning PA.allValid() or equivalent when we haven't made any change

Reply:
Done